Mathlib/CategoryTheory/Limits/Constructions/LimitsOfProductsAndEqualizers.lean

```diff
@@ -0,0 +1,480 @@
+    [HasEqualizers C] : HasFiniteLimits C where 
+  out _ := { has_limit := fun F => hasLimit_of_equalizer_and_product F }
+#align category_theory.limits.has_finite_limits_of_has_equalizers_and_finite_products CategoryTheory.Limits.hasFiniteLimits_of_hasEqualizers_and_finite_products
+
+variable {D : Type u₂} [Category.{v₂} D]
+
+/- Porting note: original file have noncomputable theory which made everything after 
+noncomputable. Removed this and made whatever necessary noncomputable -/
+-- noncomputable section
+
+section
+
+variable [HasLimitsOfShape (Discrete J) C] [HasLimitsOfShape (Discrete (Σp : J × J, p.1 ⟶ p.2)) C]
+  [HasEqualizers C]
+
+variable (G : C ⥤ D) [PreservesLimitsOfShape WalkingParallelPair G]
+  [PreservesLimitsOfShape (Discrete.{w} J) G]
+  [PreservesLimitsOfShape (Discrete.{w} (Σp : J × J, p.1 ⟶ p.2)) G]
+
+/-- If a functor preserves equalizers and the appropriate products, it preserves limits. -/
+noncomputable def preservesLimitOfPreservesEqualizersAndProduct : PreservesLimitsOfShape J G where
+  preservesLimit {K} := by
+    let P := ∏ K.obj
+    let Q := ∏ fun f : Σp : J × J, p.fst ⟶ p.snd => K.obj f.1.2
+    let s : P ⟶ Q := Pi.lift fun f => limit.π (Discrete.functor K.obj) ⟨_⟩ ≫ K.map f.2
+    let t : P ⟶ Q := Pi.lift fun f => limit.π (Discrete.functor K.obj) ⟨f.1.2⟩
+    let I := equalizer s t
+    let i : I ⟶ P := equalizer.ι s t
+    apply
+      preservesLimitOfPreservesLimitCone
+        (buildIsLimit s t (by simp) (by simp) (limit.isLimit _) (limit.isLimit _)
+          (limit.isLimit _))
+    refine' IsLimit.ofIsoLimit (buildIsLimit _ _ _ _ _ _ _) _
+    · exact Fan.mk _ fun j => G.map (Pi.π _ j)
+    · exact Fan.mk (G.obj Q) fun f => G.map (Pi.π _ f)
+    · apply G.map s
+    · apply G.map t
+    · intro f
+      dsimp [Fan.mk]
+      simp only [← G.map_comp, limit.lift_π]
+      congr 
+    · intro f
+      dsimp [Fan.mk]
+      simp only [← G.map_comp, limit.lift_π]
+      apply congrArg G.map 
+      dsimp
+    · apply Fork.ofι (G.map i)
+      rw [← G.map_comp, ← G.map_comp]
+      apply congrArg G.map
+      exact equalizer.condition s t 
+    · apply isLimitOfHasProductOfPreservesLimit
+    · apply isLimitOfHasProductOfPreservesLimit
+    · apply isLimitForkMapOfIsLimit   
+      apply equalizerIsEqualizer
+    · refine Cones.ext (Iso.refl _) ?_
+      intro j; dsimp; simp
+-- See note [dsimp, simp].
+#align category_theory.limits.preserves_limit_of_preserves_equalizers_and_product CategoryTheory.Limits.preservesLimitOfPreservesEqualizersAndProduct
+
+end
+
+/- Porting note: the original parameter [∀ (J) [Fintype J], PreservesColimitsOfShape 
+(Discrete.{0} J) G] triggered the error "invalid parametric local instance, parameter 
+with type Fintype J does not have forward dependencies, type class resolution cannot 
+use this kind of local instance because it will not be able to infer a value for this 
+parameter." Changed to implicit below. -/
```

Comment:
Should the outer or the inner `[]` be changed to `{}`?
The same for the other instance of this problem lower in the file.

Reply:
I tried that. I asked a question on Zulip: [invalid parametric local instance](https://leanprover.zulipchat.com/#narrow/stream/270676-lean4/topic/invalid.20parametric.20local.20instance)

I also tried to refactor adding a new class `PreservesFiniteProducts` but that was too involved for the port (IMO). So I expanded the porting note. This (and the colimit version) only pop each once more for left exact functors. If that is a bear, I'll come back and fiddle with this more.

Reply:
Ok. So I went back and did it as minimally as possible. I think this is the optimal solution.